[clang] Compute accurate begin location for CallExpr with explicit object parameter

HighCommander4 opened the pull request:
The explicit object parameter is written before the callee expression, so the begin location should come from the explicit object parameter.

Fixes https://github.com/llvm/llvm-project/issues/116335

zwuis commented on clang/lib/AST/Expr.cpp:
```diff
@@ -1639,6 +1639,13 @@ SourceLocation CallExpr::getBeginLoc() const {
     return OCE->getBeginLoc();
 
+  if (const CXXMethodDecl *Method =
+          dyn_cast_or_null<const CXXMethodDecl>(getCalleeDecl());
```
```suggestion
          dyn_cast_if_present<const CXXMethodDecl>(getCalleeDecl());
```

zyn0217 reviewed (commented):
I think this also needs a selection tree test to demonstrate the original case works with the current approach.

Besides, If we can't highlight the loci differences (compared to the behavior before the last PR) using simple `-ast-dump` flags, then perhaps we should reconsider the necessity of clang/test/AST/ast-dump-cxx2b-deducing-this.cpp - removing it might be a good option.

That said, I was still thinking if there's any possibility for us to find a better AST model for such use cases, rather than relying on special-casing it in `CallExpr::getBeginLoc`

HighCommander4 commented:
> That said, I was still thinking if there's any possibility for us to find a better AST model for such use cases, rather than relying on special-casing it in `CallExpr::getBeginLoc`

A more general implementation of `CallExpr::getBeginLoc` that occurs to me is: return the earlier of the callee's begin loc and the first argument's (if present) begin loc.

tbaederr commented on clang/lib/AST/Expr.cpp:
```diff
@@ -1639,6 +1639,13 @@ SourceLocation CallExpr::getBeginLoc() const {
   if (const auto *OCE = dyn_cast<CXXOperatorCallExpr>(this))
     return OCE->getBeginLoc();
 
+  if (const CXXMethodDecl *Method =
```
```suggestion
  if (const auto *Method =
```

HighCommander4 commented:
Updated to add a clangd test that fails with the previous fix approach.

> A more general implementation of `CallExpr::getBeginLoc` that occurs to me is: return the earlier of the callee's begin loc and the first argument's (if present) begin loc.

I did not do this because accurately determining which of two `SourceLocation`s is earlier requires calling `SourceManager::isBeforeInTranslationUnit()`, and a `SourceManager` does not seem to be available in `CallExpr::getBeginLoc()`.

HighCommander4 committed:
[clang] Compute accurate begin location for CallExpr with explicit object parameter

The explicit object parameter is written before the callee expression,
so the begin location should come from the explicit object parameter.

HighCommander4 commented:
(Removed extraneous formatting changes)

zyn0217 commented on clang/lib/AST/Expr.cpp:
```diff
@@ -1639,11 +1639,19 @@ SourceLocation CallExpr::getBeginLoc() const {
     return OCE->getBeginLoc();
 
+  if (const auto *Method =
+          dyn_cast_if_present<const CXXMethodDecl>(getCalleeDecl());
```
nit: I presume the `const` in `<const CXXMethodDecl>` isn't necessary, right?

cor3ntin replied:
Maybe you could do something like

```
unsigned Offset = 0;
if(isExplicitObjectMemberFunction())
    Offset = 1;
```
To avoid some duplication

HighCommander4 replied:
Sorry, I'm not quite following this suggestion. What am I supposed to do with `Offset`?

cor3ntin replied:
heh I think there is some opportunities to merge the slightly different if statements (for the explicit and not explicit cases) but it would be a minor improvement so I'll will approve that

HighCommander4 replied:
> heh I think there is some opportunities to merge the slightly different if statements (for the explicit and not explicit cases) but it would be a minor improvement so I'll will approve that

Ah, I see.

I considered doing something like this:

```c++
bool UseFirstArgLoc = false;
if (/* is explicit object */)
  UseFirstArgLoc = true;

SourceLocation begin = getCallee()->getBeginLoc();
if ((begin.isInvalid() || UseFirstArgLoc) && getNumArgs() > 0 && getArg(0))
  begin = getArg(0)->getBeginLoc();
return begin;
```
  
But this needlessly evaluates `getCallee()->getBeginLoc()` in the explicit case, so I opted not to do this.

zyn0217 reviewed (approved):
Thanks, while I haven’t come up with a better solution to model this situation, I think this looks good for now.

Please give @cor3ntin some time before merging, in case there are some better approaches that I might have overlooked.

nikic commented:
It looks like this causes a significant compile-time regression: https://llvm-compile-time-tracker.com/compare.php?from=2b855dd97092e2178ac5c470a804a17ec440d7e5&to=9ccde12f5eeb91152900082a2ae839e2a9702b31&stat=instructions:u (Maybe most clearly seen during clang bootstrap, where this adds 0.5% to many compilations: https://llvm-compile-time-tracker.com/compare_clang.php?from=2b855dd97092e2178ac5c470a804a17ec440d7e5&to=9ccde12f5eeb91152900082a2ae839e2a9702b31&stat=instructions%3Au&sortBy=interestingness)

Is that expected?

HighCommander4 commented:
> It looks like this causes a significant compile-time regression: https://llvm-compile-time-tracker.com/compare.php?from=2b855dd97092e2178ac5c470a804a17ec440d7e5&to=9ccde12f5eeb91152900082a2ae839e2a9702b31&stat=instructions:u (Maybe most clearly seen during clang bootstrap, where this adds 0.5% to many compilations: https://llvm-compile-time-tracker.com/compare_clang.php?from=2b855dd97092e2178ac5c470a804a17ec440d7e5&to=9ccde12f5eeb91152900082a2ae839e2a9702b31&stat=instructions%3Au&sortBy=interestingness)
> 
> Is that expected?

I think it's at least conceivable, as the patch adds an extra branch to `CallExpr::getBeginLoc()`, which is likely to be called pretty frequently.

The branch is **not** taken most of the time (it's only taken for calls to functions with an explicit object argument, which is a C++23 feature), so perhaps annotating the branch as "unlikely" is sufficient to avoid the performance regression?

HighCommander4 commented:
> The branch is **not** taken most of the time (it's only taken for calls to functions with an explicit object argument, which is a C++23 feature), so perhaps annotating the branch as "unlikely" is sufficient to avoid the performance regression?

Ah, no, that's not sufficient because it takes some work (a call to `getCalleeDecl()`) to compute the branch condition.

We could consider revising the implementation approach to optimize this better; for example, add a "uses explicit object argument" bit to `CallExprBits`?

tbaederr commented:
I didn't check if any tests fail but here's a version where `CallExpr` saves its `BeginLoc` explicitly: http://llvm-compile-time-tracker.com/compare.php?from=416e4cd332c7421b187844ac9aaf6fe28b575a7d&to=0b6e36fe460409aa59958b79766b4f64a31c97e6&stat=instructions:u

HighCommander4 commented:
> I didn't check if any tests fail but here's a version where `CallExpr` saves its `BeginLoc` explicitly: http://llvm-compile-time-tracker.com/compare.php?from=416e4cd332c7421b187844ac9aaf6fe28b575a7d&to=0b6e36fe460409aa59958b79766b4f64a31c97e6&stat=instructions:u

Yeah, this is another alternative, though I wonder if the memory tradeoff it makes (increasing the size of `CallExpr` by 4 bytes) is worth it.

shafik commented:
This PR was flagged as the cause of this crash: https://github.com/llvm/llvm-project/issues/130272
